### PR TITLE
Sync annotations to google drive after exiting

### LINKED ISSUE #1: Automatic upload of results
At the moment, the tool runs completely locally. It is beneficial for data privacy but risk loosing hours of work when one JSON file gets corrupted or lost. The problem could be solved by (optional) automatic upload of JSON file with results to the cloud. The job would be backed-up and supervisor can check annotator's job in real-time.

**Required features**
- uplad JSON with annotations to cloud after every app exit (or choose other appropriate interval)
- use optional flag to enable the feature. Default behavior would stay local to protect privacy  
- upload only JSON with results NOT THE DATA to allow workflow with sensitive data

### PR DESCRIPTION
**Pull request by Fizza-Rubab**
### What?
I have added functionality for uploading the annotations to Google Drive of the user. The annotations are uploaded only once after the user exits the program.
### Why?
A backup of the annotations can be stored in case the file on the local system gets corrupted. It can also be used by collaborators and supervisors to track progress on data annotation task.
### How?
Syncing to Google Drive uses the Google Drive API via the Pydrive wrapper. The user first has to setup the Google Drive API access through the Google Cloud Console (details mentioned in annotations.md in the docs). Once API and credentials are configured, the user can provide a flag in the script to enable syncing. The user can also provide a cloud folder id if they want the results to be saved in a specific folder.

### CODE REVIEW

File: annotate_bboxes.py [hunk 1]
```diff
@@ -7,8 +7,9 @@
 import numpy as np
 
 from src.bbox_annotator import BboxAnnotator
-from src.json_utils import increment_idx, load_annotations, save_annotations
+from src.json_utils import increment_idx, load_annotations, save_annotations, upload_annotations, authenticate_drive
 
+from pydrive.drive import GoogleDrive
```
Comment (MiraPurkrabek) [inline annotate_bboxes.py]: Consider making this import optional to allow the code running without PyDrive. I would opt for dynamic import only when args.cloud_upload is True but other options (like try/except etc.) are also possible.

**Review by MiraPurkrabek — CHANGES_REQUESTED** [review]
Hi @Fizza-Rubab ,
Thank you for the nice code. I tried the new feature and have several comments. Some might require code changes, while others are just curious questions.

1. How will we handle the sharing of user secrets when pushing to the repository? Having one JSON in the workdir and taking care not to accidentally upload it to GitHub seems too risky. I would propose adding the secrets.json (or maybe all JSONs) to .gitignore.
2. The current solution requires GDrive authentication after every app run. Would it be possible to authenticate the app just once and have Google remember the decision for at least some time?

**Comment by MiraPurkrabek** [comment]
Once the above-mentioned issues/questions are resolved, I propose to:

1. Add PyDrive as an optional requirement to both the .txt and Conda env files.
2. Change "Miroslav Purkrabek" to "PoseAnnotator contributors" in the BibTeX citation, as I will no longer be the sole contributor.

**Comment by Fizza-Rubab** [comment]
Hi @MiraPurkrabek, thank you for reviewing the code. 

1. I agree that the secrets file can be accidentally uploaded. It may be safer to add all *.json files to gitignore.
2. Thank you for pointing this out. Yes, it would be possible. I can save the access and refresh tokens in a file when the user authenticates the first time. They can then be loaded to prevent the authentication prompt from appearing until the file is deleted, or access is revoked.

I will share the updated code soon.

**Review by MiraPurkrabek — APPROVED** [review]

**Comment by MiraPurkrabek** [comment]
All checks completed, merging into the main.
Thank you @Fizza-Rubab for the new feature.

**Pull request merged**